### PR TITLE
MERLIN algorithm

### PR DESCRIPTION
**Pull request by emailweixu**
Now the merlin_algorithm_test.py can run successfully. It take 10-20 min to reach perfect solution on my machine.

### CODE REVIEW

File: alf/algorithms/merlin_algorithm.py [hunk 1]
```diff
@@ -0,0 +1,441 @@
...
+            activation_fn=tf.keras.activations.tanh,
+            name=name + "/value_net")
+
+        self._rnn = rnn
+        self._actor_net = actor_net
+
+        # TODO: add qvalue_net for predicting Q-value
+
+    def _variables(self):
+        return _collect_variables(self._rnn, self._value_net, self._actor_net)
```
Comment (bayesian) [inline alf/algorithms/merlin_algorithm.py]: _key_net should also be included.

File: alf/algorithms/merlin_algorithm.py [hunk 2]
```diff
@@ -0,0 +1,441 @@
...
+        latent_vector, kld, next_state = self.encode_step(inputs, state)
+
+        # TODO: decoder for action
+        decoder_loss = self.decode_step(latent_vector, observation)
+
+        return AlgorithmStep(
+            outputs=latent_vector,
+            state=next_state,
+            info=LossInfo(
+                loss=decoder_loss.loss + kld,
```
Comment (bayesian) [inline alf/algorithms/merlin_algorithm.py]: same here, should support weighted sum

File: alf/algorithms/merlin_algorithm.py [hunk 3]
```diff
@@ -0,0 +1,441 @@
...
+    else:
+        return tf.keras.layers.StackedRNNCells(cells)
+
+
+def _collect_variables(*modules):
+    modules = tf.nest.flatten(modules)
+    return sum([mod.variables for mod in modules], [])
+
+
+class MemoryBasedPredictor(Algorithm):
```
Comment (bayesian) [inline alf/algorithms/merlin_algorithm.py]: everything should be annotated with @gin.configurable?

File: alf/algorithms/merlin_algorithm.py [hunk 4]
```diff
@@ -0,0 +1,441 @@
...
+            fc_layer_params=(200, ),
+            activation_fn=tf.keras.activations.tanh,
+            name=name + "/actor_net")
+
+        super(MemoryBasedActor, self).__init__(
+            action_spec=action_spec,
+            train_state_spec=get_rnn_cell_state_spec(rnn),
+            action_distribution_spec=actor_net.output_spec)
+
+        self._loss = ActorCriticLoss(action_spec) if loss is None else loss
```
Comment (bayesian) [inline alf/algorithms/merlin_algorithm.py]: make ActorCriticLoss gin.configurable so it could let client enable more powerful features like td/gae

Comment (emailweixu) [reply alf/algorithms/merlin_algorithm.py]: It's already gin.configurable

File: alf/algorithms/merlin_algorithm.py (hunk 3 above)
Comment (emailweixu) [reply alf/algorithms/merlin_algorithm.py]: good idea

File: alf/algorithms/decoding_algorithm.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+
+from tf_agents.agents.tf_agent import LossInfo
+from tf_agents.networks.network import Network
+import tf_agents.specs.tensor_spec as tensor_spec
+
+from alf.algorithms.algorithm import Algorithm, AlgorithmStep
+from alf.utils.encoding_network import EncodingNetwork
+
+
+class DecodingAlgorithm(Algorithm):
```
[the remark this answers is not in the document]
Comment (emailweixu) [reply alf/algorithms/decoding_algorithm.py]: We'll consider doing that later if necessary.

File: alf/algorithms/memory.py [hunk 6]
```diff
@@ -165,7 +227,6 @@ def read(self, keys, scale=None):
                 assert scale.shape == keys.shape[:len(scale.shape)]
                 k = len(keys.shape) - len(scale.shape)
```
[the remark this answers is not in the document]
Comment (emailweixu) [reply alf/algorithms/memory.py]: Here we allow the possibility of multiple keys sharing one scale

File: alf/algorithms/merlin_algorithm.py (hunk 2 above)
Comment (emailweixu) [reply alf/algorithms/merlin_algorithm.py]: This weight will be handled by loss_weight in Algorithm

File: alf/algorithms/merlin_algorithm.py (hunk 1 above)
Comment (emailweixu) [reply alf/algorithms/merlin_algorithm.py]: good catch. But since tf.keras.Layer now derives from tf.Module, we can get rid of _variables() now.

**Pull request merged**